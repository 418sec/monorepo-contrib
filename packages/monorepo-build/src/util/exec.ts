import { execFileSync as _execSync } from 'child_process'

export const execSync = (cwd: string, command: string[]) =>
  _execSync(command[0], command.slice(1), { cwd, env: process.env })
