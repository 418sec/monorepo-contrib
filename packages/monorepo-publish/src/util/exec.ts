import { execSync as _execSync } from 'child_process'

export const execSync = (cwd: string, command: string) =>
  _execSync(command, { cwd })
