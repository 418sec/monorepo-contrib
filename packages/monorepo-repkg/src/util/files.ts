import { lstatSync, readdirSync } from 'fs'
import * as path from 'path'
import * as globby from 'globby'

export const isDirectory = (directoryPath: string) =>
  lstatSync(directoryPath).isDirectory()

export const getDirectories = (directoryPath: string) =>
  readdirSync(directoryPath)
    .map(name => path.join(directoryPath, name))
    .filter(isDirectory)

export const getFiles = (
  directory: string,
  {
    includes = ['**/*.js'],
    excludes = [
      'node_modules/**',
      '**/__tests__/**',
      'lib/**',
      '**/fixtures/**',
    ],
  } = {}
) =>
  globby.sync([...includes, ...excludes.map(s => `!${s}`)], {
    cwd: directory,
  })
