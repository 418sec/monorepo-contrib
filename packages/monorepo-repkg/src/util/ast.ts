import { readFileSync } from 'fs'
import * as babylon from 'babylon'
import {
  File,
  Node,
  Statement,
  CallExpression,
  MemberExpression,
} from 'babel-types'
import { NodePath } from 'babel-traverse'
import { PackageRename } from './packages'
import generator from 'babel-generator'
import * as prettier from 'prettier'
import * as Debug from 'debug'

const debug = Debug('monorepo-repkg')

export function getAst(filePath: string) {
  const code = readFileSync(filePath, 'utf8')
  const ast = babylon.parse(code, {
    sourceType: 'module',
    plugins: ['jsx', 'flow', 'objectRestSpread', 'classProperties'],
  })
  return { code, ast }
}

export function isRequire(path: NodePath<Node>) {
  const { type } = path
  if (type !== 'CallExpression') {
    return false
  }

  const node = path.node as CallExpression
  const callee = node.callee
  const arg = node.arguments[0]

  // Check for require(...) or require.func(...)
  if (
    !(
      ((callee.type === 'Identifier' && callee.name === 'require') ||
        (callee.type === 'MemberExpression' &&
          (callee.object as any).name === 'require')) &&
      arg
    )
  ) {
    return false
  }

  return true
}

export function rewriteRequireNode(
  node: CallExpression,
  packageAliases: PackageRename[],
  { fileName = '...' } = {}
) {
  const { arguments: args } = node
  const arg: any = args[0]

  if (arg.type === 'StringLiteral') {
    const orig = arg.value
    let changed = false
    for (const { oldName, newName } of packageAliases) {
      const temp = arg.value
      arg.value = arg.value.replace(oldName, newName)
      if (temp !== arg.value) {
        if (changed) {
          throw new Error('Unstable package replace')
        }
        changed = true

        debug(`${fileName}: ${orig} -> ${arg.value}`)
      }
    }

    return changed
  }

  // TODO: extract this to a passed in function
  // This is a safe require, it's our paths module which does not change.
  // There are other keys, too, I know are safe to ignore (e.g. json files).
  if (
    arg.name === 'packagePath' ||
    arg.name === 'appPackageJson' ||
    arg.name === 'templateDependenciesPath' ||
    (arg.type === 'CallExpression' &&
      arg.arguments[arg.arguments.length - 1] &&
      arg.arguments[arg.arguments.length - 1].value === 'package.json') ||
    (arg.object && arg.object.name === 'paths') ||
    (arg.type === 'BinaryExpression' && arg.left.value.startsWith('..'))
  ) {
    return false
  }

  throw new Error('Require must be a string literal.')
}

export function getCode(ast: File, originalCode: string) {
  let { code } = generator(
    ast,
    {
      sourceMaps: false,
      comments: true,
      retainLines: true,
      compact: false,
      concise: true,
    },
    originalCode
  )

  code = prettier.format(code, {
    singleQuote: true,
    trailingComma: 'es5',
  })

  return (
    code
      // Fix bash comments mangled by Babel module mode
      .replace('///usr/bin/env node', '#!/usr/bin/env node')
  )
}
