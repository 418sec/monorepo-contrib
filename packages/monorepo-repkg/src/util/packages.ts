import { lstatSync, readFileSync, writeFileSync, renameSync } from 'fs'
import { isDirectory, getDirectories, getFiles } from './files'
import * as path from 'path'
import { getAst, isRequire, rewriteRequireNode, getCode } from './ast'
import traverse from 'babel-traverse'
import { CallExpression } from 'babel-types'

export const getPackages = (directoryPath: string) =>
  getDirectories(directoryPath)
    .filter(directory => {
      try {
        return lstatSync(path.join(directory, 'package.json')).isFile()
      } catch (e) {
        if (e.code === 'ENOENT') {
          return false
        }
        throw e
      }
    })
    .map(directory => ({
      name: require(path.join(directory, 'package.json')).name as string,
      directory,
    }))
    .filter(({ name }) => name != null)

export type Package = { directory: string }

export type PackageRename = {
  oldName: string
  newName: string
} & Package

export const renamePackages = (packages: PackageRename[]) => {
  for (const { directory, oldName, newName } of packages) {
    const packagePath = path.resolve(directory, 'package.json')
    const json = JSON.parse(readFileSync(packagePath, 'utf8'))
    if (json.name !== oldName) {
      throw new Error(`Package name mismatch: ${json.name} !== ${oldName}`)
    }
    json.name = newName
    const pkgObjs = [
      'dependencies',
      'devDependencies',
      'peerDependencies',
      'optionalDependencies',
    ]
    for (const key of pkgObjs) {
      if (json[key] == null) {
        continue
      }
      for (const { oldName, newName } of packages) {
        if (json[key][oldName] == null) {
          continue
        }
        json[key][newName] = json[key][oldName]
        delete json[key][oldName]
      }
    }
    writeFileSync(packagePath, JSON.stringify(json, null, 2) + '\n')
  }
}

export const repackagePackages = (
  packagesDirectory: string,
  packages: PackageRename[]
) => {
  if (!isDirectory(packagesDirectory)) {
    return false
  }
  for (const { directory, oldName, newName } of packages) {
    const oldDirectory = path.join(packagesDirectory, oldName)
    if (path.relative(directory, oldDirectory) !== '') {
      throw new Error(`Path mismatch: ${directory} !== ${oldDirectory}`)
    }
    renameSync(oldDirectory, path.join(packagesDirectory, newName))
  }
  return true
}

export const rewriteRequires = (packages: PackageRename[]) => {
  for (const { directory } of packages) {
    const files = getFiles(directory)

    for (const file of files) {
      const filePath = path.join(directory, file)
      const { ast, code: origCode } = getAst(filePath)
      let changedFile = false
      traverse(ast, {
        enter(path) {
          if (!isRequire(path)) {
            return
          }

          if (
            rewriteRequireNode(path.node as CallExpression, packages, {
              fileName: file,
            })
          ) {
            changedFile = true
          }
        },
      })

      if (!changedFile) {
        continue
      }

      const outCode = getCode(ast, origCode)
      writeFileSync(filePath, outCode)
    }
  }
}
