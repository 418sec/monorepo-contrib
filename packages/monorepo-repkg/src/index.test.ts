import { repkg } from './index'
import * as tmp from 'tmp'
import { copySync, readFileSync } from 'fs-extra'
import * as path from 'path'

import { getFiles as _getFiles } from './util/files'

const getFiles = (dir: string) =>
  _getFiles(dir, {
    includes: ['**/*.*'],
  })

describe('repkg', () => {
  it('should purely repackage a repo', () => {
    const pkgs = path.resolve(__dirname, '..', 'fixtures', 'packages')
    const sources = getFiles(pkgs)

    const { name: directory } = tmp.dirSync()
    copySync(pkgs, directory, {
      errorOnExist: true,
      recursive: true,
    })

    // Capture source files
    expect(
      sources.map(name => readFileSync(path.resolve(pkgs, name), 'utf8'))
    ).toMatchSnapshot()

    repkg(directory)

    // Ensure source files left alone
    expect(
      sources.map(name => readFileSync(path.resolve(pkgs, name), 'utf8'))
    ).toMatchSnapshot()

    // Ensure target was repackaged
    expect(
      getFiles(directory).map(name =>
        readFileSync(path.resolve(directory, name), 'utf8')
      )
    ).toMatchSnapshot()
  })
})
