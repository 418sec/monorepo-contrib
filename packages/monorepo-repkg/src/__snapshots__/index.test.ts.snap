// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repkg should purely repackage a repo 1`] = `
Array [
  "require('package-2')
",
  "{
  \\"name\\": \\"package-1\\",
  \\"version\\": \\"1.0.0\\"
}
",
  "{
  \\"name\\": \\"package-2\\",
  \\"version\\": \\"1.0.0\\"
}
",
]
`;

exports[`repkg should purely repackage a repo 2`] = `
Array [
  "require('package-2')
",
  "{
  \\"name\\": \\"package-1\\",
  \\"version\\": \\"1.0.0\\"
}
",
  "{
  \\"name\\": \\"package-2\\",
  \\"version\\": \\"1.0.0\\"
}
",
]
`;

exports[`repkg should purely repackage a repo 3`] = `
Array [
  "require('package-2-dangerous');
",
  "{
  \\"name\\": \\"package-1-dangerous\\",
  \\"version\\": \\"1.0.0\\"
}
",
  "{
  \\"name\\": \\"package-2-dangerous\\",
  \\"version\\": \\"1.0.0\\"
}
",
]
`;
